File: src/characters/chef.ml
open! Core
open! Import

include Character.Make (struct
    type t =
      { kind : Kind.t
      ; alignment : Alignment.t
      }

    let id = "chef"
    let name = "Chef"
    let t = { kind = Kind.Townsfolk; alignment = Alignment.Good }
    let kind { kind; _ } = kind
    let alignment { alignment; _ } = alignment
  end)

let night_action ~player_id:pid ~night =
  if night <> 1
  then None
  else
    Some
      (if_alive
         pid
         (let%bind.Botc_exec state = Botc_exec.get_state
          and () = Botc_exec.wake pid in
          let%bind.Botc_exec () =
            if Game_state.is_poisoned state pid
            then (
              let%bind.Botc_exec ns =
                Botc_exec.narrator_pick "poisoned chef count" [ 0; 1; 2; 3 ] ~pick_count:1
              in
              let n = List.hd_exn ns in
              Botc_exec.tell pid [%string "%{n#Int} evil pairs"])
            else (
              let alive = Game_state.alive_ids state in
              let n = List.length alive in
              let count =
                List.counti alive ~f:(fun i pid ->
                  let next = List.nth_exn alive ((i + 1) mod n) in
                  Game_state.is_evil state pid && Game_state.is_evil state next)
              in
              Botc_exec.tell pid [%string "%{count#Int} evil pairs"])
          in
          Botc_exec.sleep pid))
;;

let day_action ~player_id:_ = None
let on_setup ~player_id:_ = None
let on_nominated ~player_id:_ ~nominator:_ = None
let on_executed ~player_id:_ = None
let on_night_kill ~player_id:_ = None

File: src/characters/chef.rs
//! Chef (Townsfolk): on the first night, learn how many pairs of evil players
//! are sitting next to each other.

use crate::ability::Ability;
use crate::ctx::Ctx;
use crate::ids::PlayerId;
use crate::role::{Alignment, CharacterInfo, Kind};

character!(Chef, "chef", "Chef", Kind::Townsfolk, Alignment::Good);

impl Ability for Chef {
    fn info(&self) -> CharacterInfo {
        Self::INFO
    }

    fn on_night(&self, ctx: &mut Ctx, me: PlayerId, night: u32) {
        if night != 1 || !ctx.grim.is_alive(me) {
            return;
        }
        ctx.wake(me);
        let seats: Vec<PlayerId> = ctx.grim.seats().collect();
        // Resolve each player's evil reading once (so a Recluse/Spy is judged
        // consistently), then count adjacent evil pairs around the ring.
        let evil: Vec<bool> = seats
            .iter()
            .map(|&id| ctx.registers_evil(id, "chef"))
            .collect();
        let n = seats.len();
        let mut pairs = 0i32;
        for i in 0..n {
            if evil[i] && evil[(i + 1) % n] {
                pairs += 1;
            }
        }
        let plausible: Vec<i32> = (0..=n as i32).collect();
        let shown = ctx.deliver(me, pairs, &plausible);
        ctx.reveal(me, &format!("{shown} pairs of evil players sit together"));
        ctx.sleep(me);
    }
}
